File: cpp_version/src/utils/logger.cpp
#include "utils/logger.h"
#include <iostream>
#include <chrono>
#include <iomanip>
#include <thread>

Logger& Logger::instance() {
    static Logger instance;
    return instance;
}

Logger::Logger() : current_level_(LogLevel::INFO) {}

Logger::~Logger() {
    if (log_file_.is_open()) {
        log_file_.close();
    }
}

void Logger::set_log_level(LogLevel level) {
    current_level_ = level;
}

void Logger::set_log_file(const std::string& filename) {
    std::lock_guard<std::mutex> lock(mutex_);
    if (log_file_.is_open()) {
        log_file_.close();
    }
    log_file_.open(filename, std::ios::app);
}

void Logger::debug(const std::string& message) {
    log(LogLevel::DEBUG, message);
}

void Logger::info(const std::string& message) {
    log(LogLevel::INFO, message);
}

void Logger::warning(const std::string& message) {
    log(LogLevel::WARNING, message);
}

void Logger::error(const std::string& message) {
    log(LogLevel::ERROR, message);
}

void Logger::critical(const std::string& message) {
    log(LogLevel::CRITICAL, message);
}

void Logger::log(LogLevel level, const std::string& message) {
    if (level < current_level_) {
        return;
    }

    std::lock_guard<std::mutex> lock(mutex_);

    auto now = std::chrono::system_clock::now();
    auto now_c = std::chrono::system_clock::to_time_t(now);
    auto now_ms = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()) % 1000;

    std::stringstream ss;
    ss << std::put_time(std::localtime(&now_c), "%Y-%m-%d %H:%M:%S")
       << '.' << std::setfill('0') << std::setw(3) << now_ms.count()
       << " [" << std::this_thread::get_id() << "] ";

    switch (level) {
        case LogLevel::DEBUG: ss << "DEBUG"; break;
        case LogLevel::INFO: ss << "INFO"; break;
        case LogLevel::WARNING: ss << "WARNING"; break;
        case LogLevel::ERROR: ss << "ERROR"; break;
        case LogLevel::CRITICAL: ss << "CRITICAL"; break;
    }

    ss << ": " << message;

    std::cout << ss.str() << std::endl;
    if (log_file_.is_open()) {
        log_file_ << ss.str() << std::endl;
    }
}

File: rust_version/src/utils/logger.rs
use std::fs::OpenOptions;
use std::io::Write;
use std::sync::{Arc, Mutex};
use chrono::Local;
use std::thread;

#[derive(PartialOrd, PartialEq, Clone, Copy)]
pub enum LogLevel {
    DEBUG,
    INFO,
    WARNING,
    ERROR,
    CRITICAL,
}

pub struct Logger {
    current_level: Mutex<LogLevel>,
    log_file: Mutex<Option<std::fs::File>>,
}

impl Logger {
    pub fn instance() -> Arc<Logger> {
        static INSTANCE: std::sync::OnceLock<Arc<Logger>> = std::sync::OnceLock::new();
        INSTANCE.get_or_init(|| Arc::new(Logger::new())).clone()
    }

    fn new() -> Self {
        Logger {
            current_level: Mutex::new(LogLevel::INFO),
            log_file: Mutex::new(None),
        }
    }

    pub fn set_log_level(&self, level: LogLevel) {
        *self.current_level.lock().unwrap() = level;
    }

    pub fn set_log_file(&self, filename: &str) -> std::io::Result<()> {
        let file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(filename)?;
        *self.log_file.lock().unwrap() = Some(file);
        Ok(())
    }

    pub fn debug(&self, message: &str) {
        self.log(LogLevel::DEBUG, message);
    }

    pub fn info(&self, message: &str) {
        self.log(LogLevel::INFO, message);
    }

    pub fn warning(&self, message: &str) {
        self.log(LogLevel::WARNING, message);
    }

    pub fn error(&self, message: &str) {
        self.log(LogLevel::ERROR, message);
    }

    pub fn critical(&self, message: &str) {
        self.log(LogLevel::CRITICAL, message);
    }

    fn log(&self, level: LogLevel, message: &str) {
        if level < *self.current_level.lock().unwrap() {
            return;
        }

        let now = Local::now();
        let thread_id = thread::current().id();

        let log_message = format!(
            "{}.{:03} [{:?}] {}: {}",
            now.format("%Y-%m-%d %H:%M:%S"),
            now.timestamp_subsec_millis(),
            thread_id,
            self.level_to_string(level),
            message
        );

        println!("{}", log_message);

        if let Some(file) = self.log_file.lock().unwrap().as_mut() {
            if let Err(e) = writeln!(file, "{}", log_message) {
                eprintln!("Failed to write to log file: {}", e);
            }
        }
    }

    fn level_to_string(&self, level: LogLevel) -> &'static str {
        match level {
            LogLevel::DEBUG => "DEBUG",
            LogLevel::INFO => "INFO",
            LogLevel::WARNING => "WARNING",
            LogLevel::ERROR => "ERROR",
            LogLevel::CRITICAL => "CRITICAL",
        }
    }
}